forem/forem
# Social Auth generalization refactor

## What type of PR is this? (check all applicable)


## Description

This PR takes the Generalization refactor from #7952 into it's own separate PR in order to simplify the complexity in an effort to ship new Authentication Providers.

Kudos to @rhymes for the original implementation over in #7952 of the generalization logic, this PR just cherry picks from there.

## Related Tickets & Documents

#7952

## QA Instructions, Screenshots, Recordings

This PR should have no effect in the authentication capabilities from master (GitHub + Twitter OAuth integrations). It only refactors the way Authentication Providers are added, so in order to test this I'd recommend:

1. Boot up locally with GitHub & Twitter Auth ENV variables setup
1. Check `/admin/config` inside `All Site Configuration` > `Authentication` to make sure both GitHub & Twitter providers are enabled
1. Make sure both Authentication providers continue to work as expected

## Added tests?


## Added to documentation?


## [optional] Are there any post deployment tasks we need to perform?

Must check and be sure both Authentication Providers (Twitter + GitHub) are enabled & configured correctly in `/admin/config`

## [optional] What gif best describes this PR or how it makes you feel?

![step one](https://media.giphy.com/media/109fP7pua6Osgw/giphy.gif)
